src/Command/Push/PushArtifactCommand.php

```diff
@@ -0,0 +1,284 @@
+
+/**
+ * Class PushArtifactCommand.
+ */
+class PushArtifactCommand extends PullCommandBase {
+
+  protected static $defaultName = 'push:artifact';
+
+  /**
+   * @var string
+   *
+   * Drupal project directory.
+   */
+  protected $dir;
+
+  /**
+   * @var array
+   *
+   * Composer vendor directories.
+   */
+  protected $vendorDirs;
+
+  /**
+   * @var array
+   *
+   * Composer scaffold files.
+   */
+  protected $scaffoldFiles;
+
+  /**
+   * {inheritdoc}.
+   */
+  protected function configure(): void {
+    $this->setDescription('Build and push a code artifact to a Cloud Platform environment')
+      ->addOption('dir', NULL, InputArgument::OPTIONAL, 'The directory containing the Drupal project to be pushed')
+      ->addOption('no-sanitize', NULL, InputOption::VALUE_NONE, 'Do not sanitize the build artifact')
+      ->addOption('dry-run', NULL, InputOption::VALUE_NONE, 'Do not push changes to Acquia Cloud')
+      ->acceptEnvironmentId()
+    ->setHelp('This command builds a sanitized deploy artifact by running <options=bold>composer install</>, removing sensitive files, and committing vendor directories.' . PHP_EOL . PHP_EOL
+      . 'Vendor directories and scaffold files are committed to the build artifact even if they are ignored in the source repository.' . PHP_EOL . PHP_EOL
+      . 'To run additional build or sanitization steps (e.g. <options=bold>npm install</>), add a <options=bold>post-install-cmd</> script to your <options=bold>composer.json</> file: https://getcomposer.org/doc/articles/scripts.md#command-events');
+  }
+
+  /**
+   * @param \Symfony\Component\Console\Input\InputInterface $input
+   * @param \Symfony\Component\Console\Output\OutputInterface $output
+   *
+   * @return int 0 if everything went fine, or an exit code
+   * @throws \Exception
+   */
+  protected function execute(InputInterface $input, OutputInterface $output): int {
+    // @todo change deploy strategy depending on whether this is a "source" repo (requiring building) or not
+    // @todo handle if Git user name/email is missing
+    $this->setDirAndRequireProjectCwd($input);
+    $is_dirty = $this->isLocalGitRepoDirty();
+    $commit_hash = $this->getLocalGitCommitHash();
+    if ($is_dirty) {
+      throw new AcquiaCliException('Pushing code was aborted because your local Git repository has uncommitted changes. Please either commit, reset, or stash your changes via git.');
+    }
+    $this->checklist = new Checklist($output);
+
+    // @todo handle environments with tags deployed
+    $output->writeln('<info>You must select an environment with a Git branch deployed</info>');
+    $environment = $this->determineEnvironment($input, $output, TRUE);
+    if (strpos($environment->vcs->path, 'tags') === 0) {
+      throw new AcquiaCliException("You cannot push to an environment that has a git tag deployed to it. Environment {$environment->name} has {$environment->vcs->path} deployed. Please select a different environment.");
```

Comment:
This is a possibly strange assumption. Why do we care about what is checked out on the environment? Are we pushing to remote or are we pushing to an environment? If we are pushing to a remote and it doesn’t matter. If we are pushing to an environment, why can’t we check out a different branch or tag as part of that push?

Reply:
we probably should check out the pushed branch on the remote environment. I’m fine with preventing that for environments that have tags, but it only makes sense if we are otherwise going to be potentially deploying a new branch.

Reply:
I want the user to think of deploys as being to an _environment_, not to a particular branch or tag. This was biggest point of confusion with `blt deploy`, people didn't understand that they had to deploy to _git_ first, and then deploy git to an environment.

This is easy to accomplish technically if an environment is running a branch, we just push to the branch. It's less clear what we should do if an environment runs a tag. I'm kicking that decision down the road in the spirit of getting this out as an MVP.